### PR TITLE
SQL-554: Implement MongoSQLDatabaseMetadata, MongoSQL dialect meta me…

### PR DESCRIPTION
**Pull request by bucaojit**
…thods

JIRA: https://jira.mongodb.org/browse/SQL-554
Implemented MongoSQLDatabaseMetadata, MongoSQL dialect meta methods:
getClientInfoProperties
getSQLKeywords, 
getTypeInfo, 
getTableTypes

Does not have testing included.  Similar to the PR for SQL-553 , depends on SQL-513, SQL-535, and SQL-551

### CODE REVIEW

File: src/main/java/com/mongodb/jdbc/MongoSQLDatabaseMetaData.java
Comment (bucaojit) [reply src/main/java/com/mongodb/jdbc/MongoSQLDatabaseMetaData.java]: Sounds good, using the new MongoJsonSchema methods now.

File: src/main/java/com/mongodb/jdbc/MongoSQLDatabaseMetaData.java
Comment (bucaojit) [reply src/main/java/com/mongodb/jdbc/MongoSQLDatabaseMetaData.java:33]: I updated the keywords list based on the linked [keywords document](https://developer.mimer.com/wp-content/uploads/standard-sql-reserved-words-summary.pdf)

File: src/main/java/com/mongodb/jdbc/MongoSQLDatabaseMetaData.java
Comment (bucaojit) [inline src/main/java/com/mongodb/jdbc/MongoSQLDatabaseMetaData.java:1131]: @nbagnard informed me that the clientInfoProperties are not for connection settings.  Postgres for example only returns applicationName for the session, mysql returns empty set.  So removed the connection settings and passing the empty cursor instead.  
I will keep the mysql clientInfoProperties as is to not change functionality.

**Review by pmeredit — APPROVED** [review]
LGTM once the other suggestions are met. Oh, but do make sure that the current comment for keywords matches what we have now.

File: src/main/java/com/mongodb/jdbc/MongoDatabaseMetaData.java
Comment (mattChiaravalloti) [inline src/main/java/com/mongodb/jdbc/MongoDatabaseMetaData.java:1143]: [organizational nit] My initial vision for this these helper methods was to make them `static`, possibly even part of a totally separate class and `public static`. What if we create a `BsonTypeInfo` class (or some other relevant name) that has all of these as `static` methods? We can also move the `BSON_<type>_TYPE_NAME` constants to that class.

This feels like a big chunk of related fields and methods that could be factored out separately. They are pretty specific to `MongoDatabaseMetaData` subtypes, but still it could be nice to have a larger number of smaller classes than continuing to build these large monolithic classes. What do you think?

File: src/main/java/com/mongodb/jdbc/MongoDatabaseMetaData.java
Comment (mattChiaravalloti) [inline src/main/java/com/mongodb/jdbc/MongoDatabaseMetaData.java:1178]:
```suggestion
    protected boolean getCaseSensitivity(String typeName) throws SQLException {
```
Why return `Integer` and `Boolean` from these methods instead of their primitive counterparts?

**Review by mattChiaravalloti — CHANGES_REQUESTED** [review]
This is looking very good. We're almost there 🏁  I just have two questions about the new type info helper methods.

File: src/main/java/com/mongodb/jdbc/MongoDatabaseMetaData.java
Comment (nbagnard) [inline src/main/java/com/mongodb/jdbc/MongoDatabaseMetaData.java:1211]: I thought that fixedPrecScale was for currency/money values, which are generally related to Decimal/numeric with a fixed precision and scale. I don't know if reporting Int and Long will cause issues or not, but I don't think they are great type for representing money values either. Maybe we should just return false for every type?

File: src/main/java/com/mongodb/jdbc/MongoDatabaseMetaData.java
Comment (nbagnard) [inline src/main/java/com/mongodb/jdbc/MongoDatabaseMetaData.java:1239]: Shouldn't min scale be 0 for Decimal and Double too? For Decimal, I'm pretty sure that you can have a scale < 34 and 0 is how you would represent an int/long value.

File: src/main/java/com/mongodb/jdbc/MongoDatabaseMetaData.java
Comment (nbagnard) [inline src/main/java/com/mongodb/jdbc/MongoDatabaseMetaData.java:1241]: For Double, it's a floating point decimal so I am not totally sure because it does not really apply, but certainly it can be < 15. I think 0 also makes sense.

File: src/main/java/com/mongodb/jdbc/MongoDatabaseMetaData.java
Comment (nbagnard) [inline src/main/java/com/mongodb/jdbc/MongoDatabaseMetaData.java:1271]: @pmeredit , @mattChiaravalloti : is this true? 
Also, I'm still not sure if we should report a max scale for a floating point decimal.

File: src/main/java/com/mongodb/jdbc/MongoDatabaseMetaData.java
Comment (nbagnard) [inline src/main/java/com/mongodb/jdbc/MongoDatabaseMetaData.java:1277]: Based on this page, I think the max scale for BSON Date is 3, not 0 : https://docs.mongodb.com/manual/reference/mongodb-extended-json/#mongodb-bsontype-Date
Maybe @pmeredit or @mattChiaravalloti you can confirm/infirm?

Comment (mattChiaravalloti) [reply src/main/java/com/mongodb/jdbc/MongoDatabaseMetaData.java:1277]: Agreed that the docs do seem to say that!

File: src/main/java/com/mongodb/jdbc/MongoDatabaseMetaData.java
Comment (nbagnard) [reply src/main/java/com/mongodb/jdbc/MongoDatabaseMetaData.java:1239]: Got the answer from conversation in standup : BSON_DECIMAL scale is always 34.

File: src/main/java/com/mongodb/jdbc/MongoDatabaseMetaData.java
Comment (mattChiaravalloti) [reply src/main/java/com/mongodb/jdbc/MongoDatabaseMetaData.java:1160]: I agree that `null` is the right move here, given the documentation's description.

File: src/main/java/com/mongodb/jdbc/MongoDatabaseMetaData.java
Comment (bucaojit) [reply src/main/java/com/mongodb/jdbc/MongoDatabaseMetaData.java:1178]: You're right, primitives are more appropriate.  I'll make the change.

File: src/main/java/com/mongodb/jdbc/MongoDatabaseMetaData.java
Comment (bucaojit) [reply src/main/java/com/mongodb/jdbc/MongoDatabaseMetaData.java]: Okay that makes sense.  I moved the helper functions and the `BSON_<type>_TYPE_NAME` to a new class: `BsonTypeInfo`

File: src/main/java/com/mongodb/jdbc/MongoDatabaseMetaData.java
Comment (bucaojit) [reply src/main/java/com/mongodb/jdbc/MongoDatabaseMetaData.java]: Checked with @nbagnard and good to resolve.

File: src/main/java/com/mongodb/jdbc/MongoDatabaseMetaData.java
Comment (bucaojit) [reply src/main/java/com/mongodb/jdbc/MongoDatabaseMetaData.java]: Okay, setting these `0` and `unknownLength` to `null` as mentioned in an earlier comment.

File: src/main/java/com/mongodb/jdbc/MongoDatabaseMetaData.java
Comment (bucaojit) [reply src/main/java/com/mongodb/jdbc/MongoDatabaseMetaData.java]: I created a class that will help with this: `BsonTypeInfo`.  I'll create a refactoring ticket to make better use of `BsonTypeInfo` in other classes besides MongoDatabaseMetadata.

File: src/main/java/com/mongodb/jdbc/MongoDatabaseMetaData.java
Comment (bucaojit) [reply src/main/java/com/mongodb/jdbc/MongoDatabaseMetaData.java]: Okay, I am fine with this.  I'll make the change.

File: src/main/java/com/mongodb/jdbc/MongoDatabaseMetaData.java
Comment (bucaojit) [reply src/main/java/com/mongodb/jdbc/MongoDatabaseMetaData.java]: I switched getPrecision() back to Integer in order to return `null`.

File: src/main/java/com/mongodb/jdbc/MongoDatabaseMetaData.java
Comment (mattChiaravalloti) [reply src/main/java/com/mongodb/jdbc/MongoDatabaseMetaData.java]: awesome, thanks

File: src/main/java/com/mongodb/jdbc/BsonTypeInfo.java
Comment (mattChiaravalloti) [inline src/main/java/com/mongodb/jdbc/BsonTypeInfo.java:6]:
```suggestion
    public static final String BINDATA_TYPE_NAME = "binData";
```
[ultra-nit] Totally up to you (or anyone else who has a strong opinion): we could remove the BSON prefix from all of these since they'll be accessed using the class name (`BsonTypeInfo.XYZ_TYPE_NAME`).

File: src/main/java/com/mongodb/jdbc/MongoSQLDatabaseMetaData.java
Comment (mattChiaravalloti) [inline src/main/java/com/mongodb/jdbc/MongoSQLDatabaseMetaData.java:369]: should we add a `static getJDBCType` method to `BsonTypeInfo` to get this for each type? I know I also need this info in at least one other place

**Review by mattChiaravalloti — APPROVED** [review]
Looks good to me! 🦅 

I have two small suggestions/requests that are up to you. The helper method I suggested would be nice to see but I'll leave that up to you. No need to retag when/if you make changes!

File: src/main/java/com/mongodb/jdbc/MongoSQLDatabaseMetaData.java
Comment (bucaojit) [reply src/main/java/com/mongodb/jdbc/MongoSQLDatabaseMetaData.java:369]: I think we can defer this to the refactor change Natacha is suggesting with making BsonTypeInfo into enums.  The enum would have the TYPE_NAME, DATA_TYPE specified then we can call `enumType.getJdbcType()` such as:
```
       for (BsonTypeInfo currType : BsonTypeInfo.values()) {
           if (currType.isMySqlSupported) {
               docs.add(
                       MySQLDatabaseMetaData.getTypeInfoValuesDoc(
                               currType.getTypeName(),
                               currType.getJdbcType(), //dataType
                               currType.getPrecision(), //precision
                               ... ));
           }
       }
```

**Pull request merged**
